pkg/prom/cluster/node.go

```diff
@@ -0,0 +1,327 @@
+			return fmt.Errorf("failed to stop ring: %w", err)
+		}
+	}
+
+	if !cfg.Enabled {
+		n.cfg = cfg
+		return nil
+	}
+
+	r, err := newRing(cfg.Lifecycler.RingConfig, "agent_viewer", agentKey, n.reg)
+	if err != nil {
+		return fmt.Errorf("failed to create ring: %w", err)
+	}
+	if err := n.reg.Register(r); err != nil {
+		return fmt.Errorf("failed to register ring metrics: %w", err)
+	}
+	if err := services.StartAndAwaitRunning(context.Background(), r); err != nil {
+		return fmt.Errorf("failed to start ring: %w", err)
+	}
+	n.ring = r
+
+	lc, err := ring.NewLifecycler(cfg.Lifecycler, n, "agent", agentKey, false, n.reg)
+	if err != nil {
+		return fmt.Errorf("failed to create lifecycler: %w", err)
+	}
+	if err := services.StartAndAwaitRunning(context.Background(), lc); err != nil {
+		r.StopAsync()
+		return fmt.Errorf("failed to start lifecycler: %w", err)
+	}
+	n.lc = lc
+
+	n.cfg = cfg
+
+	// Reload and reshard the cluster.
+	n.reload <- struct{}{}
+	return nil
+}
+
+// WaitJoined waits for the node the join the cluster and enter the
+// ACTIVE state.
+func (n *node) WaitJoined(ctx context.Context) error {
+	n.mut.RLock()
+	defer n.mut.RUnlock()
+
+	level.Info(n.log).Log("msg", "waiting for the node to join the cluster")
+	defer level.Info(n.log).Log("msg", "node has joined the cluster")
+
+	return waitJoined(ctx, agentKey, n.ring.KVClient, n.lc.ID)
+}
+
+func waitJoined(ctx context.Context, key string, kvClient kv.Client, id string) error {
+	kvClient.WatchKey(ctx, key, func(value interface{}) bool {
+		if value == nil {
+			return true
+		}
+
+		desc := value.(*ring.Desc)
+		for ingID, ing := range desc.Ingesters {
+			if ingID == id && ing.State == ring.ACTIVE {
+				return false
+			}
+		}
+
+		return true
+	})
+
+	return ctx.Err()
+}
+
+// newRing creates a new Cortex Ring that ignores unhealthy nodes.
+func newRing(cfg ring.Config, name, key string, reg prometheus.Registerer) (*ring.Ring, error) {
+	codec := ring.GetCodec()
+	store, err := kv.NewClient(
+		cfg.KVStore,
+		codec,
+		kv.RegistererWithKVName(reg, name+"-ring"),
+	)
+	if err != nil {
+		return nil, err
+	}
+	return ring.NewWithStoreClientAndStrategy(cfg, name, key, store, ring.NewIgnoreUnhealthyInstancesReplicationStrategy())
+}
+
+// run waits for connection to the ring and kickstarts the join process.
+func (n *node) run() {
+	for range n.reload {
+		n.mut.RLock()
+
+		if err := n.performClusterReshard(context.Background(), true); err != nil {
+			level.Warn(n.log).Log("msg", "dynamic cluster reshard did not succeed", "err", err)
+		}
+
+		n.mut.RUnlock()
+	}
+
+	level.Info(n.log).Log("msg", "node run loop exiting")
+}
+
+// performClusterReshard informs the cluster to immediately trigger a reshard
+// of their workloads. if includeSelf is true, the server provided to newNode will
+// also be informed. includeSelf should be true when joining the cluster, and false
+// when leaving.
+func (n *node) performClusterReshard(ctx context.Context, includeSelf bool) error {
```

Comment:
Thoughts on `includeSelf` renaming to `isJoiningCluster` or similar. Feels more descriptive to the comment.